lib/midcom/helper/imagepopup/handler/upload.php

```diff
@@ -0,0 +1,159 @@
+<?php
+/**
+ * @author lukasz chalat
+ * @package midcom.helper.imagepopup
+ * @copyright The Midgard Project, http://www.midgard-project.org
+ * @license http://www.gnu.org/licenses/lgpl.html GNU Lesser General Public License
+ */
+
+/**
+ * This handler get uploaded file and save it in database and write to file.
+ *
+ * @package midcom.helper.imagepopup
+ */
+class midcom_helper_imagepopup_handler_upload extends midcom_baseclasses_components_handler
+{
+    /**
+     *
+     * @param mixed $handler_id The ID of the handler.
+     * @param array $args The argument list.
+     * @param array &$data The local request data.
+     */
+    public function _handler_upload($handler_id, array $args, array &$data)
+    {
+        // Get the file
+        reset($_FILES);
+        $temp = array_shift($_FILES);
+        
+        $accepted_origin = midcom::get()->get_host_name();
+        
+        if(is_uploaded_file($temp['tmp_name']))
+        {
+            if(isset($_SERVER['HTTP_ORIGIN']))
+            {
+                // same-origin requests won't set an origin. If the origin is set, it must be valid.
+                if(in_array($_SERVER['HTTP_ORIGIN'], $accepted_origin))
+                {
+                    midcom::get()->header('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+                } 
+                else
+                {
+                    throw new midcom_error_forbidden("HTTP/1.0 403 Origin Denied");
+                }
+            }
+            
+            if(preg_match("/([^\w\s\d\-_~,;:\[\]\(\).])|([\.]{2,})/", $temp['name']))
```

Comment:
Do we even need this test? `midcom_db_attachment::safe_filename()` should make any input filename url-safe, and besides, files here are always modified versions of already existing attachments, right? so theoretically they should already have a valid filname (I think, should be tested probably)